Plugins: 'channel_opened' notification type

darosior opened the pull request:
This adds a new notification type : `channelopened`.

This notification is emitted if a peer successfully opened a channel to us, and sends the id of the peer, the funding amount, and the feerate.

ZmnSCPxj commented on tests/plugins/misc_notifications.py:
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env python3
+"""Plugin to be used to test miscellaneous notifications.
+
+Only used for 'channel_opened' for now.
```
`channelopened` is what is actually used.

ZmnSCPxj commented on tests/test_plugin.py:
```diff
@@ -476,3 +476,15 @@ def test_warning_notification(node_factory):
+    l1, l2 = node_factory.line_graph(2, fundchannel=True, fundamount=amount,
+                                     opts=opts)
+    l2.daemon.wait_for_log(r"A channel to us was opened by {}, with an amount"
+                           " of {}msat*".format(l1.info["id"], amount*1000))
```
Triggers a warning:

>      tests/test_plugin.py:490:70: E226 missing whitespace around arithmetic operator

darosior commented:
Thanks for the review @ZmnSCPxj, I corrected and will resolve conflict in the upcoming rebase.

cdecker commented on doc/PLUGINS.md:
```diff
@@ -200,6 +200,20 @@ corresponding payloads are listed below.
 
+#### `channelopened`
+
+A notification for topic `channelopened` is sent if a peer successfully funded a channel
```
We seem to get inconsistent with the naming of hooks:

 - `htlc_accepted`
 - `invoice_payment`
 - `db_write`
 - `channelopened`

I'd suggest adding the `_` for consistency's sake :-)

darosior replied:
Ok :-)

darosior commented:
Corrected the naming and the include duplication.

cdecker commented on tests/test_plugin.py:
```diff
@@ -494,3 +494,15 @@ def test_invoice_payment_notification(node_factory):
+
+def test_channel_opened_notification(node_factory):
+    """
+    Test the 'channelopened' notification sent at channel funding success.
```
```suggestion
    Test the 'channel_opened' notification sent at channel funding success.
```

cdecker commented on tests/plugins/misc_notifications.py:
```diff
@@ -0,0 +1,20 @@
+
+
+@plugin.subscribe("channel_opened")
+def on_channelopened(plugin, channel_opened):
```
```suggestion
def on_channel_opened(plugin, channel_opened):
```

cdecker commented on lightningd/notification.c:
```diff
@@ -76,3 +77,17 @@ void notify_invoice_payment(struct lightningd *ld, struct amount_msat amount,
+			 struct amount_sat *funding_sat, u32 *feerate_per_kw)
+{
+	struct jsonrpc_notification *n =
+	    jsonrpc_notification_start(NULL, notification_topics[4]);
```
The use of indices here can lead to unwanted renamings if the `notification_topics` array get reordered. Prefer adding an inline string here.

ZmnSCPxj replied:
Probably appropriate to modify all of the notifications in a separate commit, then, as there is existing code which uses the same pattern.

darosior commented:
Arf, forgot these ones

cdecker commented on doc/PLUGINS.md:
```diff
@@ -200,6 +200,20 @@ corresponding payloads are listed below.
+	"channel_opened": {
+		"id": "03864ef025fde8fb587d989186ce6a4a186895ee44a926bfc370e2c366597a3f8f",
+		"funding_satoshis": "100000000msat",
+		"feerate": 7500,
```
`feerate` seems like an odd detail to add here, but something like the funding txid would be useful here imho.

darosior replied:
Ok, I'll replace the feerate by the txid

darosior commented:
Done :)

darosior commented:
> Actual commit messages use `channelopened` and `openchannel`, should use `channel_opened`?

Yes, I had not the time to change it again yesterday.

While including the `funding_txid` as requested by @cdecker I moved the notification to the `opening_fundee_finished()` function, which (I think) makes more sense. I also added a drive-by commit to remove the use of `notification_topics` indices in notification functions. The notification now sends :
```
{
	"channel_opened": {
		"id": "03864ef025fde8fb587d989186ce6a4a186895ee44a926bfc370e2c366597a3f8f",
		"funding_satoshis": "100000000msat",
		"funding_txid": "4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b",
		"funding_locked": false
	}
}
```

ZmnSCPxj commented on lightningd/opening_control.c:
```diff
@@ -550,7 +550,7 @@ static void opening_funder_finished(struct subd *openingd, const u8 *resp,
 	derive_channel_id(&fc->cid, &channel->funding_txid, funding_txout);
 
-	/* Was this an external wallet initiation ? */
+		/* Was this an external wallet initiation ? */
```
Unnecessary change?

darosior commented:
Corrected, thanks.

cdecker commented:
ACK https://github.com/ElementsProject/lightning/pull/2840/commits/1221981e36b7357aa78e81f7bb9eb3a03bca9b9a
